### ISSUE TITLE
Destroying via nested attributes on has_one does not clear in-memory association

### ISSUE THREAD

**Issue by 12eh** [rails/rails#31990]
### Steps to reproduce
Delete an associated object via `_destroy` in nested attributes, on a `has_one` association. That association is still non-null on the parent object until you call `reload`.

runnable test case: https://gist.github.com/12eh/8dff5d9a1ad4b173f4e5c40a405d1d38

### Expected behavior
The association should return `nil`, consistent with other modifications made by `update` and how nested attributes behave with `has_many`.

### Actual behavior
The deleted object is still present.

### System configuration
**Rails version**: 5.2.0

**Ruby version**: 2.5.1p57

**Comment by victorhazbun** [comment]
Can you please test of the latest stable version of Rails? I'm interested in making a fix if needed on the latest stable version (today).

**Comment by 12eh** [comment]
@victorhazbun sorry for the delayed reply; I removed `, github: "rails/rails"` from the linked test script and re-ran. The assertion on line 52 still fails, and the output before then states "Using rails 5.2.0".

**Comment by FreekingDean** [comment]
Looks like the tests are somewhat explicitly not testing this:
https://github.com/rails/rails/blob/a1ac186/activerecord/test/cases/nested_attributes_test.rb#L312

**Comment by 12eh** [comment]
Perhaps, but I'd put emphasis on the "somewhat" -- that `reload` could easily be left over from a copy/pasted test case, rather than being intentionally added. In either case, my two cents is that the current behavior is not intuitive, because it's the only case I can think of where `post.update(...)` does not update the in-memory data accessible from `post`.

For the second point, I'm not sure I follow. When destroying the child via the parent object's association (`post.comment.destroy!`), it _does_ clear it in memory. If you're talking about a slight variation on your posted test case of `comment.destroy!; refute_nil post.comment`, I don't think that's analogous to the case I posted -- obviously AR is not going to keep track of and update all variables/relations pointing to the thing that just got deleted, but I do expect it to update the one which I used to perform the deletion.
